path_login.go

```diff
@@ -333,7 +339,17 @@ func (b *kubeAuthBackend) parseAndValidateJWT(jwtStr string, role *roleStorageEn
 	}
 
 	// verify the namespace is allowed
-	if len(role.ServiceAccountNamespaces) > 1 || role.ServiceAccountNamespaces[0] != "*" {
+	valid := false
+	if role.ServiceAccountNamespaceSelector != "" {
+		if valid, err = b.namespaceValidatorFactory(config).validateLabels(ctx,
+			client, sa.namespace(), role.ServiceAccountNamespaceSelector); err != nil {
+			return nil, err
+		}
+	}
+	if !valid && len(role.ServiceAccountNamespaces) == 0 {
+		return nil, logical.CodedError(http.StatusForbidden, "namespace not authorized")
+	}
+	if !valid && (len(role.ServiceAccountNamespaces) > 1 || role.ServiceAccountNamespaces[0] != "*") {
```

Comment:
I am bit confused by these `if !valid` conditionals, I wonder if they belong under the conditional at line 343, since valid will always be false if `role.ServiceAccount...Selector` is empty string? I am guessing  valid only makes sense in the context of namespace label validation?

Reply:
I will restructure this in a more readable way.